components/core/src/clp/ir/LogEventSerializer.cpp

```diff
@@ -0,0 +1,168 @@
+#include "LogEventSerializer.hpp"
+
+#include <string_utils/string_utils.hpp>
+
+#include "../ffi/ir_stream/encoding_methods.hpp"
+#include "../ffi/ir_stream/protocol_constants.hpp"
+#include "spdlog_with_specializations.hpp"
+
+using std::string;
+using std::string_view;
+
+namespace clp::ir {
+template <typename encoded_variable_t>
+LogEventSerializer<encoded_variable_t>::~LogEventSerializer() {
+    if (true == m_is_open) {
+        SPDLOG_ERROR("Serializer is not closed before being destroyed - output maybe corrupted");
+    }
+}
+
+template <typename encoded_variable_t>
+auto LogEventSerializer<encoded_variable_t>::open(
+        string const& file_path,
+        epoch_time_ms_t reference_timestamp
+) -> ErrorCode {
+    static_assert(std::is_same_v<encoded_variable_t, four_byte_encoded_variable_t>);
+
+    m_serialized_size = 0;
+    m_num_log_events = 0;
+    m_ir_buffer.clear();
+    m_prev_msg_timestamp = reference_timestamp;
+
+    m_writer.open(file_path, FileWriter::OpenMode::CREATE_FOR_WRITING);
+    m_zstd_compressor.open(m_writer);
+    // For now, use preset metadata for the preamble
+    if (false
+        == clp::ffi::ir_stream::four_byte_encoding::serialize_preamble(
+                TIMESTAMP_PATTERN,
+                TIMESTAMP_PATTERN_SYNTAX,
+                TIME_ZONE_ID,
+                reference_timestamp,
+                m_ir_buffer
+        ))
+    {
+        SPDLOG_ERROR("Failed to serialize preamble");
+        m_zstd_compressor.close();
+        m_writer.close();
+        return ErrorCode_Failure;
+    }
+    m_is_open = true;
+
+    // Flush the preamble
+    flush();
+
+    return ErrorCode_Success;
+}
+
+template <typename encoded_variable_t>
+auto LogEventSerializer<encoded_variable_t>::open(string const& file_path) -> ErrorCode {
+    static_assert(std::is_same_v<encoded_variable_t, eight_byte_encoded_variable_t>);
+
+    m_serialized_size = 0;
+    m_num_log_events = 0;
+    m_ir_buffer.clear();
+
+    m_writer.open(file_path, FileWriter::OpenMode::CREATE_FOR_WRITING);
+    m_zstd_compressor.open(m_writer);
+
+    // For now, use preset metadata for the preamble
+    if (false
+        == clp::ffi::ir_stream::eight_byte_encoding::serialize_preamble(
+                TIMESTAMP_PATTERN,
+                TIMESTAMP_PATTERN_SYNTAX,
+                TIME_ZONE_ID,
+                m_ir_buffer
+        ))
+    {
+        SPDLOG_ERROR("Failed to serialize preamble");
+        m_zstd_compressor.close();
+        m_writer.close();
+        return ErrorCode_Failure;
+    }
+    m_is_open = true;
+
+    // Flush the preamble
+    flush();
+
+    return ErrorCode_Success;
+}
+
+template <typename encoded_variable_t>
+auto LogEventSerializer<encoded_variable_t>::close() -> void {
+    m_ir_buffer.push_back(clp::ffi::ir_stream::cProtocol::Eof);
+    flush();
+    m_zstd_compressor.close();
+    m_writer.close();
```

Comment:
This is trivial, but I think we could make a helper to wrap these two methods so we never mess up the order for closing them